meson.build

```diff
@@ -0,0 +1,180 @@
+# SPDX-FileCopyrightText: 2022 Aetf <aetf@unlimited-code.works>
+#
+# SPDX-License-Identifier: MIT
+
+project('kmscon', 'c',
+  version: '9.0.0',
+  license: 'MIT',
+  # meson 0.58: f-string
+  # meson 0.62: dependency libdl
+  meson_version: '>=0.62.0',
+  default_options: [
+    'warning_level=1',
+    'werror=true',
+    'buildtype=debugoptimized',
+    'c_std=gnu99',
+  ]
+)
+
+add_project_arguments(
+  '-ffast-math',
+  '-fno-strict-aliasing',
+  '-ffunction-sections',
+  '-fdata-sections',
+  '-fstack-protector',
+  '-fvisibility=hidden',
+  '-D_GNU_SOURCE',
+  '-D_POSIX_C_SOURCE=200809L',
+  language: 'c'
+)
+
+#
+# Directory structure
+#
+prefix = get_option('prefix')
+bindir = get_option('bindir')
+sysconfdir = get_option('sysconfdir') / meson.project_name()
+libexecdir = get_option('libexecdir') / meson.project_name()
+mandir = get_option('mandir')
+moduledir = get_option('libdir') / meson.project_name()
+
+#
+# Required dependencies
+#
+python = import('python').find_installation()
+fs = import('fs')
+
+xkbcommon_deps = dependency('xkbcommon', version: '>=0.5.0')
+libtsm_deps = dependency('libtsm', version: '>=4.0.0')
+libudev_deps = dependency('libudev', version: '>=172')
+libdrm_deps = dependency('libdrm')
+dl_deps = dependency('dl')
+threads_deps = dependency('threads')
+
+#
+# Dependencies
+# This unconditionally checks for all dependencies even if they are disabled. We
+# later look whether all required depedencies are met and finish the
+# configuration.
```

Comment:
Why? Meson has better logging if you feed the feature option directly to the dependency() lookup. Also it plays better with subproject fallbacks (for dependencies that use Meson too).

Reply:
Could you elaborate on how the overall logic works when feeding the feature option directly to `dependency()`?
I just started using Meson so I might miss something very obvious.

AFAIK, an `auto` feature is equivalent to `required: false` in `dependency()`, and then to know whether to build with the feature, I need to check if all dependencies are found.
This feels less elegant because the information is duplicated twice (one time with `dependency()` calls, the other time when checking all dependencies for a feature).

Also,  how does Meson `dependency()` handle multiple features requiring the same dependency? For example, both `renderer_gltex` and `video_drm3d` wants `glesv2`.

Reply:
When dependencies are missing for features that are enabled, `dependency()` can raise an error message that tells you the dependency was not found, this information is lost in the current .require().allowed() approach.

When dependencies aren't wanted for features that are disabled, meson will go faster by not even looking for them, and log "Dependency foo skipped: feature bar disabled".

You're checking whether it got found anyway, so it's not entirely clear to me why you find it inelegant?

Reply:
I came across [this](https://github.com/mesonbuild/meson/issues/7013#issuecomment-617505762) which I believe is exactly the problem we are talking about here, and it also happens to be your comment years ago :)

So I assume that is what you are talking about. And my point of inelegant is that when, for example, the dependency requirement for `option_tool_a` changes, you'll need to update two places:

* The `dependency()` calls that go from feature options to dependency objects
* The `if` statements for `tool_a`

With my current approach, the dependencies for a feature is specified once in an array, and later code only checks for the `enable_feature_a` variable, and doesn't need to care about what dependencies are.
(I'm not arguing that mine is the best. I just try to understand what is the best practice and why.)

Regarding two features requiring one dependency, I find mesonbuild/meson#5206 and mesonbuild/meson#8512, but it's not clear to me how those utility methods can be used to combine feature options to pass to the `dependency()` call.